# 命名空间与作用域
## 命名空间（namespace）
* **Python的命名空间是一个字典，字典内保存了变量名称与对象之间的映射关系。**
* 一个变量名称可以定义在多个不同的命名空间下，之间互不影响
* 同一个命名空间下不存在有两个相同的变量名
* 顾名思义，namespace理解为为变量名称分组划分的空间区域

#### 一般有三种命名空间
* 内置名称：built-in names,python内置的名称，如函数名abs，sum,异常名Exception等
* 全局名称：global names, python模块中定义的名称，如模块中的类，函数，模块级别的变量与常量
* 局部名称：local names, 函数中定义的名称，函数内部声明的变量名称

## 作用域
**作用域是python程序可以直接访问的命名空间的文本区域。**
* 在python程序中，直接访问一个变量，会从内到外依次访问所有的作用域直到找到，LEGB
* L:local,当前函数内部作用域
* E:Enclosing Functions,外部嵌套函数的作用域
* G:global,全局作用域
* B:built-in,python内建(内置模块)作用域

```python
g_count = 0  # 全局作用域 G
def outer():
    o_count = 1  # 闭包函数外的函数中 E
    def inner():
        i_count = 2  # 局部作用域 L
```

#### 查看python的内置名称
```python
import builtins
# dir():返回模块的属性列表
print(dir(builtins))

['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EnvironmentError', 'Exception', 'False', 'FileExistsError', 'FileNotFoundError', 'FloatingPointError', 'FutureWarning', 'GeneratorExit', 'IOError', 'ImportError', 'ImportWarning', 'IndentationError', 'IndexError', 'InterruptedError', 'IsADirectoryError', 'KeyError', 'KeyboardInterrupt', 'LookupError', 'MemoryError', 'ModuleNotFoundError', 'NameError', 'None', 'NotADirectoryError', 'NotImplemented', 'NotImplementedError', 'OSError', 'OverflowError', 'PendingDeprecationWarning', 'PermissionError', 'ProcessLookupError', 'RecursionError', 'ReferenceError', 'ResourceWarning', 'RuntimeError', 'RuntimeWarning', 'StopAsyncIteration', 'StopIteration', 'SyntaxError', 'SyntaxWarning', 'SystemError', 'SystemExit', 'TabError', 'TimeoutError', 'True', 'TypeError', 'UnboundLocalError', 'UnicodeDecodeError', 'UnicodeEncodeError', 'UnicodeError', 'UnicodeTranslateError', 'UnicodeWarning', 'UserWarning', 'ValueError', 'Warning', 'WindowsError', 'ZeroDivisionError', '__build_class__', '__debug__', '__doc__', '__import__', '__loader__', '__name__', '__package__', '__spec__', 'abs', 'all', 'any', 'ascii', 'bin', 'bool', 'bytearray', 'bytes', 'callable', 'chr', 'classmethod', 'compile', 'complex', 'copyright', 'credits', 'delattr', 'dict', 'dir', 'divmod', 'enumerate', 'eval', 'exec', 'exit', 'filter', 'float', 'format', 'frozenset', 'getattr', 'globals', 'hasattr', 'hash', 'help', 'hex', 'id', 'input', 'int', 'isinstance', 'issubclass', 'iter', 'len', 'license', 'list', 'locals', 'map', 'max', 'memoryview', 'min', 'next', 'object', 'oct', 'open', 'ord', 'pow', 'print', 'property', 'quit', 'range', 'repr', 'reversed', 'round', 'set', 'setattr', 'slice', 'sorted', 'staticmethod', 'str', 'sum', 'super', 'tuple', 'type', 'vars', 'zip']

```

#### 内部作用域要修改外部作用域的变量，需要用到global,nonlocal
具体参看[变量的作用域与闭包](变量的作用域.md)

* 函数内部修改全局变量,用global

```python
def func():
    a = 99  # 此时a只是函数内部的定义局部变量，与全局变量a没有任何关系

if __name__ == '__main__':
    a = 1
    func()
    print(a)  # 1
```

```python
def func():
    # 此时python在函数内部作用域中没找到对a的定义
    # 然后去全局空间找到了a
    print(a)  # 1

if __name__ == '__main__':
    a = 1
    func()
    print(a)  # 1
```

```python
def func():
    # 该代码无法编译通过
    # 此时认为a是函数内部的局部变量，但是在函数内部并没有找到对a的定义声明
    print(a)
    a = 99

if __name__ == '__main__':
    a = 1
    func()
    print(a)
```

```python
def func():
    # 声明a是引用的全局变量a，不是函数局部变量
    global a
    a = 99

if __name__ == '__main__':
    a = 1
    func()
    print(a)  # 99
```

* 嵌套函数内函数修改外函数中的变量，用 nonlocal

```python
def out_():
    a = 1

    def in_():
        nonlocal a
        a = 99

    in_()
    print(a)  # 99

if __name__ == '__main__':
    out_()
```
